ggshield/core/git_shell.py

```diff
@@ -76,8 +85,20 @@ def check_git_dir(wd: Optional[str] = None) -> None:
         raise UsageError("Not a git directory.")
 
 
+@lru_cache(None)
```

Comment:
We should not cache the return of this function because if `wd` is None the function uses the current working directory, making it dependent of outside parameters. For example something like this:

```python
os.chdir("foo")
foo_root = get_git_root()
os.chdir("bar")
bar_root = get_git_root()  # error: get_git_root() returns the root for foo
```

This is admittedly unlikely to happen right now, and changing the current working directory in a running process is not a good idea but:
- we have been bitten by that in the past, better not add another time-bomb;
- part of our code does change the current working directory 😞.

This implementation also causes `git rev-parse --git-dir`  and `git rev-parse --show-toplevel` to be called multiple times, which can be slow on Windows (because Windows is slow at spawning processes). A way to avoid this would be to lru_cache the results of the calls. Something like this:

```python
@lru_cache
def _git_rev_parse(option: str, wd: str) -> Optional[str]:
    try:
        return git(["rev-parse", option], cwd=wd)
    except subprocess.CalledProcessError:
        return None
```

Then it can be used by the other functions:

```python
def is_git_dir(wd: str) -> bool:
    return _git_rev_parse("--git-dir", wd) is not None

def is_git_working_tree(wd: str) -> bool:
    return _git_rev_parse("--show-toplevel", wd) is not None

def get_git_root(wd: Optional[str] = None) -> str:
    if wd is None:
        wd = os.getcwd()
    check_git_dir(wd)
    top_level = _git_rev_parse("--show-toplevel", wd)
    if top_level is not None:
        return top_level
    return str(Path(_git_rev_parse("--git-dir", cwd=wd)).resolve())

```